### PR TITLE
Reduce strictness of directory layout spec-equality check

### LINKED ISSUE #21839: SHA-1 prefix error when trying to build HDF5
### Steps to reproduce the issue

I'm trying to build HDF5 with nvhpc 21.2 on a cluster. I've passed the flags target=x86_64 while building my packages. I'm also using perl 5.30.3 (hash: **f6wiogo**) since the latest version in incompatible with nvhpc.

```console
$ spack install hdf5 +hl +cxx +fortran target=x86_64 %nvhpc ^/f6wiogo
```

### Error Message

```console
Error: Specs ncurses@6.2%nvhpc@21.2~symlinks+termlib arch=linux-centos7-x86_64 and ncurses@6.2%nvhpc@21.2~symlinks+termlib arch=linux-centos7-x86_64 ^pkgconf@1.7.3%nvhpc@21.2 patches=6e94c68d45d4cb8148c8c53699921d10b13784f758f0563ce64c838e0ca48ceb arch=linux-centos7-x86_64 have the same SHA-1 prefix!
```

Here is the input spec and it's concretized form
```console
Input spec
--------------------------------
hdf5%nvhpc+cxx+fortran+hl arch=linux-None-x86_64
    ^perl@5.30.3%nvhpc@21.2+cpanm+shared+threads patches=d11fd1545df35266bda5cf67f4ed4babcd21caebb27db4146f14d8bb98a1cec4 arch=linux-centos7-x86_64
        ^berkeley-db@18.1.40%nvhpc@21.2 arch=linux-centos7-x86_64
        ^gdbm@1.18.1%nvhpc@21.2 arch=linux-centos7-x86_64
            ^readline@8.0%nvhpc@21.2 arch=linux-centos7-x86_64
                ^ncurses@6.2%nvhpc@21.2~symlinks+termlib arch=linux-centos7-x86_64

Concretized
--------------------------------
hdf5@1.10.7%nvhpc@21.2+cxx~debug+fortran+hl~java+mpi+pic+shared~szip~threadsafe api=none arch=linux-centos7-x86_64
    ^numactl@2.0.14%nvhpc@21.2 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94 arch=linux-centos7-x86_64
        ^autoconf@2.69%nvhpc@21.2 arch=linux-centos7-x86_64
            ^m4@1.4.18%nvhpc@21.2+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,5746cf51f45b405661c3edae7a78c33d41e54d83f635d16e2bf1f956dbfbf635,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-centos7-x86_64
                ^libsigsegv@2.12%nvhpc@21.2 arch=linux-centos7-x86_64
            ^perl@5.30.3%nvhpc@21.2+cpanm+shared+threads patches=d11fd1545df35266bda5cf67f4ed4babcd21caebb27db4146f14d8bb98a1cec4 arch=linux-centos7-x86_64
                ^berkeley-db@18.1.40%nvhpc@21.2 arch=linux-centos7-x86_64
                ^gdbm@1.18.1%nvhpc@21.2 arch=linux-centos7-x86_64
                    ^readline@8.0%nvhpc@21.2 arch=linux-centos7-x86_64
                        ^ncurses@6.2%nvhpc@21.2~symlinks+termlib arch=linux-centos7-x86_64
        ^automake@1.16.3%nvhpc@21.2 arch=linux-centos7-x86_64
        ^libtool@2.4.6%nvhpc@21.2 arch=linux-centos7-x86_64
    ^openmpi@4.0.5%nvhpc@21.2~atomics~cuda~cxx~cxx_exceptions+gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~pmi~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=none patches=fba0d3a784a9723338722b48024a22bb32f6a951db841a4e9f08930a93f41d7a schedulers=none arch=linux-centos7-x86_64
        ^hwloc@2.4.0%nvhpc@21.2~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-centos7-x86_64
            ^libpciaccess@0.16%nvhpc@21.2 patches=6e08dc445ece06e9e8b1344397f2d3f169005703ddc0f2ae24f366cde78c7377 arch=linux-centos7-x86_64
                ^pkgconf@1.7.3%nvhpc@21.2 patches=6e94c68d45d4cb8148c8c53699921d10b13784f758f0563ce64c838e0ca48ceb arch=linux-centos7-x86_64
                ^util-macros@1.19.1%nvhpc@21.2 arch=linux-centos7-x86_64
            ^libxml2@2.9.10%nvhpc@21.2~python patches=05ff238cf435825ef835c7ae39376b52dc83d8caf19e962f0766c841386a305a,10a88ad47f9797cf7cf2d7d07241f665a3b6d1f31fa026728c8c2ae93e1664e9 arch=linux-centos7-x86_64
                ^libiconv@1.16%nvhpc@21.2 arch=linux-centos7-x86_64
                ^xz@5.2.5%nvhpc@21.2~pic arch=linux-centos7-x86_64
                ^zlib@1.2.11%nvhpc@21.2+optimize+pic+shared arch=linux-centos7-x86_64
        ^libevent@2.1.12%nvhpc@21.2+openssl arch=linux-centos7-x86_64
            ^openssl@1.1.1j%nvhpc@21.2+systemcerts arch=linux-centos7-x86_64
```

### Information on your system

* **Spack**: 0.16.0-1384-26d2356
* **Python**: 3.6.8
* **Platform**: linux-centos7-nehalem
* **Concretizer**: original

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by scheibelp**
fixes https://github.com/spack/spack/issues/21839

https://github.com/spack/spack/pull/21735 increased the amount of spec data written (including build deps) which may trigger arguably-erroneous spec-hash collision errors. This reduces the strictness of the spec comparison check.

### CODE REVIEW

File: lib/spack/spack/directory_layout.py [hunk 1]
```diff
@@ -347,7 +347,13 @@ def check_installed(self, spec):
         #
         # TODO: remove this when we do better concretization and don't
         # ignore build-only deps in hashes.
-        elif installed_spec == spec.copy(deps=('link', 'run')):
+        elif (installed_spec.copy(deps=('link', 'run')) ==
+              spec.copy(deps=('link', 'run'))):
```
Comment (alalazo) [inline lib/spack/spack/directory_layout.py:351]: There are a few things that are not clear to me at a first glance:
- Is this related to #21491? 
- Why is the comment mentioning `full_hash`? I think `__eq__` compares the `_cmp_key` value which is different from the full hash. Am I missing something?

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/directory_layout.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/directory_layout.py:351]: I don't think this is related to #21491. The issue here is that installed specs now contain references to their build deps, even though we don't store them separately based on those. So we need to strip those deps out before testing whether two specs are equivalent *for the purpose of whether to reinstall*, even if they aren't identical including build deps.
